### PR TITLE
Add chapter on Solver and fix minor style issues

### PR DESCRIPTION
**Pull request by yanchith**
@janper Is this satisfactory or did you envision something else? Also please note the PR is draft, because there are still TODOs in there for you to renumber the table of contents and re-generate the HTML docs.

### CODE REVIEW

**Comment by janper** [comment]
could you also please explain:
1. how the solver turns non canonical set of slots into canonical
2. what is observation and that it is random
3. how the change triggered by an observation propagates

**Comment by yanchith** [comment]
Sure I can, but 2) and 3) are vanilla WFC. If we are not counting on the reader already knowing this, then we have way more to explain throughout the entire document. If that's the case, I'd start with a chapter 0 describing vanilla WFC. I think I still do have that text lying around.

**Comment by janper** [comment]
i don't like leaving blank spots, so I'd go for explaining as much as we can with as simple words as possible so that even we architects can understand

**Comment by yanchith** [comment]
Ok, I agree, but then I would really start with a chapter called "What is Wave Function Collapse" somewhere early on. perhaps as 1.3 or 1.4 and explain WFC to the best of our ability there.

Then in the solver chapter, I'd address point 1), because that is our extension.

**Comment by janper** [comment]
sure. if you don't want to do it completely on your own, you can assign certain paragraphs to me

**Comment by yanchith** [comment]
I have already written that text last summer, we just didn't use it. Going to repurpose that.

**Comment by janper** [comment]
please also review of it it digestible for non programmers (check with Subdigital?). the new text is on the verge of it

**Comment by yanchith** [comment]
The text was already reviewed by at least @kr1sta2 and you :) I'll ask her to go through it again and post it on our internal channels.

**Comment by yanchith** [comment]
The last commit reuses some of my old text to explain vanilla WFC. I am still going to explain point 1) you mentioned above in the solver chapter.

File: README.md [hunk 1]
```diff
@@ -90,16 +93,76 @@
...
+
+## 1.4. Wave Function Collapse
+
+Before we delve deeper, let's shortly explain the WFC algorithm itself. Note
+that Monoceros extends some concepts over the original WFC, but it is very
+helpful to understand the simplified version of the algorithm first. To that
+end, it might also be helpful to watch the excellent [explanatory talk by Oskar
+Stålberg at EPC2018](https://www.youtube.com/watch?v=0bcZb-SsnrA).
+
+Wave Function Collapse (WFC) is an procedural generation algorithm that
```
Comment (janper) [inline README.md:111]: a procedural 

generative (?)

File: README.md [hunk 2]
```diff
@@ -90,16 +93,76 @@
...
+essentially attempts to satisfy constraints on a world (see
+[CSP](https://en.wikipedia.org/wiki/Constraint_satisfaction_problem) for
+more). Its name is inspired by quantum mechanics, but the similarity is just at
+the surface level - WFC is runs completely deterministically.
```
Comment (janper) [inline README.md:115]: -is

File: README.md [hunk 3]
```diff
@@ -90,16 +93,76 @@
...
+  graph](https://en.wikipedia.org/wiki/Graph_(discrete_mathematics)), in our
+  case represented as an orthogonal 3D grid.
+
+- A set of Rules describing what can occupy each tile on the grid depending on
+  what its neighbors are.
+
+The grid tiles are graph nodes, also called Slots in WFC (and in
+Monoceros). Graph edges are implicitly derived from slot neighborhood: two Slots
+are connected by an edge, if they are adjacent to each other on the grid. Every
+Slot contains a list of Modules (conceptually geometry in case of Monoceros)
```
Comment (janper) [inline README.md:130]: not sure they are a geometry. the bond between a module and its geometry is rather loose. the module parts are defined by points inside them, not the geometry stored in the module. 
i've been struggling with explaining this concept also in other parts of the docu, so maybe you can help me out somehow?

File: README.md [hunk 4]
```diff
@@ -90,16 +93,76 @@
...
+    3. **Constraint Progagation:** Remove Modules from neighboring Slots based
+       on the Rules. Repeat recursively in [depth-first
+       order](https://en.wikipedia.org/wiki/Depth-first_search) for each Slot
+       modified this way.
+
+3. If WFC generated a contradictory result, start over again with a different
+   random state.
+
+WFC does not necessarily always find a solution. If (and how easily) a valid
+solution is computed very much depends on the set of provided Rules. There are
```
Comment (janper) [inline README.md:156]: make bold

File: README.md [hunk 5]
```diff
@@ -306,6 +369,63 @@ A Typed Rule preview can be baked.
 
 #### 1.6.3.3. Indifferent Typed Rule
 
+
+## 1.7. Solver
```
Comment (janper) [inline README.md:373]: I'd make this peer to 1.4. Wave Function Collapse. i know this can't come before the three paragraphs on the data types, but it would be great if this started where 1.4 left off and started with a statement, that Monoceros WFC Solver (which should also be the heading of this paragraph) is an extension of the vanilla WFC (which as a term should be mentioned in 1.4) that differs in the following aspects...

File: README.md [hunk 6]
```diff
@@ -306,6 +369,63 @@ A Typed Rule preview can be baked.
...
+
+TODO(jp): Renumber titles and table of contents
+
+TODO(jp): Generate HTML docs
+
+Over the course of multiple iterations (called observations), Wave Function
+Collapse gradually changes the state of Slots from non-deterministic (allowing
+multiple Modules) to deterministic (allowing exactly one Module). This iterative
+process happens in the Solver component, where the Slots are observed guided by
```
Comment (janper) [inline README.md:382]: "observed guided" sounds strange

File: README.md [hunk 7]
```diff
@@ -306,6 +369,63 @@ A Typed Rule preview can be baked.
...
 #### 1.6.3.3. Indifferent Typed Rule
 
+
+## 1.7. Solver
+
+TODO(jp): Renumber titles and table of contents
+
+TODO(jp): Generate HTML docs
+
+Over the course of multiple iterations (called observations), Wave Function
```
Comment (janper) [inline README.md:379]: shouldn't this be a paragraph after 1.7.1 Canonical World State?

File: README.md [hunk 8]
```diff
@@ -306,6 +369,63 @@ A Typed Rule preview can be baked.
...
+Collapse gradually changes the state of Slots from non-deterministic (allowing
+multiple Modules) to deterministic (allowing exactly one Module). This iterative
+process happens in the Solver component, where the Slots are observed guided by
+seeded pseudo-random numbers until either every Slot ends up in a deterministic
+state (success), or any Slot ends up in a contradictory state (failure). If the
+result is contradictory, the Solver component internally re-tries up to a
+predefined number of attempts, each attempt using the already modified PRGN
```
Comment (janper) [inline README.md:386]: PRGN?

File: README.md [hunk 9]
```diff
@@ -306,6 +369,63 @@ A Typed Rule preview can be baked.
...
+Wave Function Collapse as described by [Stålberg's EPC2018
+talk](https://www.youtube.com/watch?v=0bcZb-SsnrA) a Module always has the
+dimensions to occupy exactly one Slot. While Monoceros Modules are aligned to
+the discrete grid, they can span multiple Slots and be of any voxel-based shape
+as long as their Parts connect in a continuous fashion (i.e. they touch with
+faces and not just edges). A Monoceros Module Part (not available as public data
+type) is equivalent to vanilla WFC Module.
+
+For clearer separation of concerns, this is internally implemented with
+desugaring and reconstruction on the C# side of the language boundary. In the
```
Comment (janper) [inline README.md:401]: shoudln't we call the C# side something like "Grasshopper wrapper" or so?

File: README.md [hunk 10]
```diff
@@ -306,6 +369,63 @@ A Typed Rule preview can be baked.
...
+  constraint propagation phases on an already Canonical world.
+
+For example, runnning an observation without subsequently propagating does not
+produce a Canonical world and WFC does not define how this world should behave
+if observed or propagated any further.
+
+Setting the state of the world manually rarely produces a Canonical
+world. Therefore the Rust solver always Canonicalizes the world before running
+an observe/propagate operation. It also notifies the Grasshopper solver
+component that this happened, so that it can in turn notify the user.
```
Comment (janper) [inline README.md:427]: Not anymore. It wasn't useful for the user.

What is useful is what step solver will do and we can mention it already here - the canonization of the world removes invalid neighborhoods, leaving a world in a state, that indicates which module can possibly end up where. indeed we have not exposed this feature yet but we can mention we'll include it in the coming releases.

**Review by janper — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (yanchith) [reply README.md:130]: Ok, I would reword it to mention that a Module could be anything and that it is the user of the algorithm that assigns semantics to it, and I'll also list some examples of what people might want to assign, geometry being one of them.

Can you assign anything other than geometry to Modules in Monoceros? Speaking in broader contexts, other things that are usually assigned there are: images, navigation/collision geometry, descriptions (e.g. this is a park vs this is a skyscraper). Thoughts?

File: README.md (hunk 2 above)
Comment (yanchith) [reply README.md:115]: Not sure I follow, which "is" would you like to remove?

(Btw github now has "suggestions", so if you see a mechanical change, you can just produce a diff to correct it in the github UI. It will create a review comment the PR author can just apply directly from the github UI)

File: README.md (hunk 3 above)
Comment (janper) [reply README.md:130]: not directly. i was even hesitating with the geometry because you can deconstruct a Slot to get the inscribed Module name and with that you can do anything, including placing the geometry from a source different than the Module. but then it seemed too advanced for a general user to do. that is the only reason why the geometry is still a property of a Module

File: README.md (hunk 2 above)
Comment (janper) [reply README.md:115]: WFC is runs

File: README.md (hunk 8 above)
Comment (yanchith) [reply README.md:386]: Good catch. I meant PRNG. Or should I spell that out?

Comment (janper) [reply README.md:386]: i know it is obvious from the context but still, I'd go for the full wording

File: README.md (hunk 7 above)
Comment (yanchith) [reply README.md:379]: Depending on what we want to achieve with the chapter. I viewed the 1.7.1. Canonical World State more as a tangent, while the 1.7. more of a technical description of what is going on inside the Solver component, where the canonicalization is just one interesting thing that's happening. I can maybe change the paragraph about desugaring into its own chapter as well and extend it a bit.

File: README.md (hunk 9 above)
Comment (yanchith) [reply README.md:401]: Whatever you prefer. I don't particularly like the word wrapper, so I tried to avoid it. I'd suggest "C# Grasshopper component" or "C# Grasshopper code"

File: README.md (hunk 10 above)
Comment (yanchith) [reply README.md:427]: Ok, I'll reword it :)

File: README.md (hunk 7 above)
Comment (janper) [reply README.md:379]: i though it would be more chronological

File: README.md (hunk 8 above)
Comment (yanchith) [reply README.md:386]: Used "random state" in the end :)

File: README.md (hunk 5 above)
Comment (yanchith) [reply README.md:373]: Yeah, they are related, but while 1.4. Wave Function Collapse is relevant to everyone, the 1.7. Monoceros WFC Solver is useful mostly for people wanting to understand the implementation, so it can come later in the docs. I'd do the following:

- Introduce the term "vanilla WFC" already in 1.4.
- Link back to 1.4. from 1.7. to restore the reader's context.

File: README.md (hunk 3 above)
Comment (yanchith) [reply README.md]: Ah, okay. See what I did in the text. I mentioned that a Module could be anything, but that geometry is a common usecase

File: README.md (hunk 7 above)
Comment (yanchith) [reply README.md:385]: Please have a look if it works better after my last change.

**Review by janper — APPROVED** [review]

**Comment by yanchith** [comment]
@janper If you are happy, I am happy. Merge when ready. You might also want to generate the HTML docs (not sure how you do that) so they are in sync.
